pkg/validator/app.go

```diff
@@ -82,9 +114,44 @@ func validateSelectedRoles(
 			errorMessages = append(errorMessages, invalidMsg)
 		}
 	}
+	return strings.Join(errorMessages, "\n")
+}
 
-	if len(errorMessages) == 0 {
-		return ""
+// validateRoles checks each role for property constraints not expressable
+// in the schema. Currently this just means checking that the role must
+// specify an image if there is no top-level default image.
+func validateRoles(
+	appCR *kdv1.KubeDirectorApp,
+) string {
+
+	for _, role := range appCR.Spec.NodeRoles {
+		if role.Image.RepoTag == "" {
+			if appCR.Spec.Image.RepoTag == "" {
+				return noDefaultImage
```

Comment:
to be consistent with the other uniqueness validating functions,
you may want to use the "var errorMessages []string" and "return strings.Join(errorMessages, "\n")"
construct eventhough there will only ever be at most 1 string in the string array.

Reply:
It might also be the right time for me to introduce a little more infrastructure around the validation functions. I'm not sure I want to go hog wild with that just yet, but at least: if each of those functions is eseentially going to generate a slice of error messages, then that's what each such function should return. The caller can finally compile all of those into a single string (or whatever it wants to do).

Let me see how much that will churn things.